runtime/common/src/parachains.rs

```diff
@@ -818,6 +970,117 @@ pub fn ensure_parachain<OuterOrigin>(o: OuterOrigin) -> result::Result<ParaId, B
 	}
 }
 
+
+/// Ensure that double vote reports are only processed if valid.
+#[derive(Encode, Decode, Clone, Eq, PartialEq)]
+pub struct ValidateDoubleVoteReports<T: Trait + Send + Sync>(rstd::marker::PhantomData<T>) where
+	<T as system::Trait>::Call: IsSubType<Module<T>, T>;
+
+impl<T: Trait + Send + Sync> rstd::fmt::Debug for ValidateDoubleVoteReports<T> where
+	<T as system::Trait>::Call: IsSubType<Module<T>, T>
+{
+	fn fmt(&self, f: &mut rstd::fmt::Formatter) -> rstd::fmt::Result {
+		write!(f, "ValidateDoubleVoteReports<T>")
+	}
+}
+
+/// Custom validity error used while validating double vote reports.
+#[repr(u8)]
+pub enum DoubleVoteValidityError {
+	/// The authority being reported is not in the authority set.
+	NotAnAuthority = 0,
+
+	/// Failed to convert offender's `FullIdentificationOf`.
+	FailedToConvertId = 1,
+
+	/// The signature on one or both of the statements in the report is wrong.
+	InvalidSignature = 2,
+
+	/// The two statements in the report are not conflicting.
+	NotDoubleVote = 3,
+}
+
+impl<T: Trait + Send + Sync> SignedExtension for ValidateDoubleVoteReports<T> where
+	<T as system::Trait>::Call: IsSubType<Module<T>, T>
+{
+	const IDENTIFIER: &'static str = "ValidateDoubleVoteReports";
+	type AccountId = T::AccountId;
+	type Call = <T as system::Trait>::Call;
+	type AdditionalSigned = ();
+	type Pre = ();
+	type DispatchInfo = DispatchInfo;
+
+	fn additional_signed(&self)
+		-> rstd::result::Result<Self::AdditionalSigned, TransactionValidityError>
+	{
+		Ok(())
+	}
+
+	fn validate(
+		&self,
+		_who: &Self::AccountId,
+		call: &Self::Call,
+		_info: DispatchInfo,
+		_len: usize,
+	) -> TransactionValidity {
+		let r = ValidTransaction::default();
+
+		if let Some(local_call) = call.is_sub_type() {
+			if let Call::report_double_vote(report) = local_call {
+				let e = TransactionValidityError::from(
+					InvalidTransaction::Custom(DoubleVoteValidityError::NotAnAuthority as u8)
+				);
+
+				let validators = <session::Module<T>>::validators();
+				let parent_hash = <system::Module<T>>::block_hash(<system::Module<T>>::block_number());
+
+				let authorities = Module::<T>::authorities();
+				let offender_idx = match authorities.iter().position(|a| *a == report.identity) {
+					Some(idx) => idx,
+					None => return Err(e),
+				};
+
+				let e = TransactionValidityError::from(
+					InvalidTransaction::Custom(DoubleVoteValidityError::FailedToConvertId as u8)
+				);
+				if T::FullIdentificationOf::convert(validators[offender_idx].clone()).is_none() {
+					return Err(e);
+				}
+
+				let e = TransactionValidityError::from(
+					InvalidTransaction::Custom(DoubleVoteValidityError::InvalidSignature as u8)
+				);
+				report.verify(parent_hash).map_err(|_| e)?;
+
+				let e = TransactionValidityError::from(
```

Comment:
This pre-emptive `let e = ...` pattern and then `return Err(e)` is kind of weird compared to `return Err(...)`. Mostly because `e` is rebound several times and exists in a scope where it is not relevant.